### PR TITLE
[FEAT]: Screenshot Farm - Save and Share Image 📷

### PR DESCRIPTION
**Pull request by Pranav-yadav**
# Description

## Screenshot Farm - Save and Share Image 📷
This PR add functionality,
Take screenshot of your farm _(only visible area)_ and either **save** it or **tweet** it using buttons provided in modal which also shows preview of the screenshot.

Refer Screenshot (demo video) 👇

## Includes
- **`ScreenshotButton`** RFC
- **Screenshot Button** _placed just above the audio player_
- **Modal** _shows preview of the screenshot taken_
- **Save** Button _to save the screenshot locally on device_
- **Tweet** Button _to tweet the screenshot; uses twitter intent :)_
  _(screenshot is not attached to the tweet through twitter intent because of twitter intent restrictions.
  Drag the preview image from modal and drop it into tweet intent opened in another window, as shown in below demo)_

## Screenshot (demo)
Screenshot-Button-Demo-SFL
image

Fixes # (-)

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

- `yarn test`
- Manual testing

# Checklist:

- [x] Title of the PR is relevant and is prefixed with [FEAT], [CHORE] or [FIX]
- [x] Screenshot if it includes any UI changes
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by j923sneaks** [comment]
cool! a few questions:

1. does this screenshot the whole map, including empty regions?
2. is the image quality acceptable?

im asking because i also used html2canvas but i find my pics blurry. also, you can ignore the hud div completely using `ignoreElements` of `Html2CanvasOptions`. it goes something like this:

Hud.tsx
```js
<div id="hud">
  <Menu captureFarm={captureFarm} />
  <Balance/>
  <Inventory/>
  <AudioPlayer/>
  <Address />
</div>
```

html2canvas
```js
const canvas = await html2canvas(document.body, { 
  ignoreElements: (el: Element) => el.id === 'hud',
});
```

Perhaps you can use this with `data-html2canvas-ignore` to ignore more elements. Though I'm wondering if it's needed for components not displayed such as `OuterPanel`?

i'll share my code [here](https://github.com/j923sneaks/sunflower-land/commit/822274b5f053c5fab64a4c56df087ef1c2fe837f) just in case. this approach limits the screenshot to just the farm premises.

if you want to render your component inside `ScreenshotButton.tsx`, you can just pass `gameboard` ref or query it from document body.

Hope this helps, at least to start discussions :)

**Comment by Pranav-yadav** [comment]
> 1. does this screenshot the whole map, including empty regions?

Currently only the viewport window. Working on full map ss, rn. :)



> 2. is the image quality acceptable?
> 
> im asking because i also used html2canvas but i find my pics blurry.

Depends as, if u take the ss of full screen (window) it will be definitely viewed by under lesser aspect ratio on social networks.
Yep, the images are blurry because the `CanvasRenderingContext2D` interface from `Canvas API` does the rendering and it does  **not** renders the **`pixelated-images`** as we're using `img {image-rendering: pixelated;}`, as well as **`border-images`**

Also, thanks for pointing out we can just ignore complete `Hud` itself :)
Optimized the code.

**Comment by j923sneaks** [comment]
> Currently only the viewport window. Working on full map ss, rn. :)

Is it necessary to screenshot the whole map though?

> Yep, the images are blurry because the CanvasRenderingContext2D interface from Canvas API does the rendering and it does not renders the pixelated-images as we're using img {image-rendering: pixelated;}, as well as border-images

Cool. Thanks for the explanation! Unfortunate that we can't improve image quality. This was the main thing for me continuing this feature (when I started, at least).

Once you're done, im planning to add an option to screenshot the farm premises, if it's not in there yet 😄

**Review by adamhannigan — COMMENTED** [review]
Nice work @Pranav-yadav 

The screenshot is a bit blurry but I think it will pass for the first implementation 👏 

I noticed that the `html2canvas` package has an unpacked size of 3MB. Luckily when bundled it is extremely small (must be some good tree shaking). 

The only suggestion I would make is removing the `package-lock.json`. We use `yarn` as our package manager 🤗

File: src/features/hud/components/ScreenshotButton.tsx [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+export const ScreenshotButton: React.FC = () => {
+  const [isOpen, setIsOpen] = useState(false);
+  const [ssImg, setSSImg] = useState("");
+
+  const handleTweetClick = () => {
+    window.open(
+      "https://twitter.com/intent/tweet?text=Checkout%20my%20Sunflower%20Land%20Farm%3A%0A%0A&ref_src=https%3A%2F%2Fsunflower-land.com&hashtags=SunflowerLand",
+      "_blank"
+    );
+    // setIsOpen(false);
```
Comment (Rofrtd) [inline src/features/hud/components/ScreenshotButton.tsx:18]: commented code

File: src/features/hud/components/ScreenshotButton.tsx [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    document.body.appendChild(link);
+    link.click();
+    document.body.removeChild(link);
+  };
+
+  const handleSaveImage = () =>
+    downloadImage(clearUrl(ssImg), "My SFL Farm", "jpeg");
+
+  function getScreenshot(): void {
+    //, {scale: 1,width: 1600,height: 1600, }
```
Comment (Rofrtd) [inline src/features/hud/components/ScreenshotButton.tsx:42]: commented code

File: src/features/hud/components/ScreenshotButton.tsx [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    html2canvas(document.body)
+      .then(async function (canvas: HTMLCanvasElement): Promise<string> {
+        const img: string = await canvas.toDataURL("image/jpeg");
+        return img;
+      })
+      .then((img: string) => {
+        setSSImg(img);
+        setIsOpen(true);
+        // window.open("", "_blank")?.document.write('<img src="' + img + '" />');
```
Comment (Rofrtd) [inline src/features/hud/components/ScreenshotButton.tsx:51]: commented code

File: src/features/hud/components/ScreenshotButton.tsx [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+      });
+  }
+
+  return (
+    <div className="position-fixed flex bottom-44 p-1 right-2 w-10 z-50 cursor-pointer sm:w-12 md:w-18 lg:w-fit">
+      <Button onClick={() => getScreenshot()} className="p-0">
+        <img src={screenshotIcon} className="w-fit" alt="screenshot" />
+        {/* img: w-8 mb-1 sm:w-6 sm:h-5 */}
```
Comment (Rofrtd) [inline src/features/hud/components/ScreenshotButton.tsx:59]: commented code

**Review by Rofrtd — COMMENTED** [review]
Nice work. 

Interesting how the NFTs look better than the map
image

As @adamhannigan mentioned we use `yarn` so you could remove the `package-lock.json`

File: src/features/hud/components/ScreenshotButton.tsx (hunk 2 above)
Comment (Pranav-yadav) [reply src/features/hud/components/ScreenshotButton.tsx:42]: I kept them intentionally as this feat itself is in beta. 
Removed the comments. 👍

**Review by adamhannigan — APPROVED** [review]
LGTM 🚀

**Comment by Pranav-yadav** [comment]
> The only suggestion I would make is removing the `package-lock.json`. We use `yarn` as our package manager 🤗

Yeah @adamhannigan , it got committed when I merged the SFL main in feat branch and resolved some conflicts.
Removed it.
Tho, should we add entry for it in `.gitignore` as, lot of dev's use both npm and yarn ?

**Pull request merged**